File: metadata-learning/create-conversation-graph.js
const MongoClient = require('mongodb').MongoClient;
const assert = require('assert');
const Twit = require('twit');
const env = require('./.env');


var graph = {};

async function connectToDB(dbUrl, dbName) {
    return (await MongoClient.connect(dbUrl)).db(dbName);
}

async function cursorToArray(cursor) {
    let docs = [];
    for (let doc = await cursor.next(); doc != null; doc = await cursor.next()) {
        docs.push(doc);
    }
    return docs;
}


async function prepareGraph(twitter_id, db) {
    console.log('Preparing graph for', twitter_id);
    let allThere = await cursorToArray(
        await db.collection('tweets').find({'user.screen_name': twitter_id,
                                            'in_reply_to_status_id_str': {$ne: null}
                                           }
                                          )
    );

    for (let each of allThere) {
        console.log(each.id_str);
        graph[each.id_str] = [each];
        let taskqueue = [each.in_reply_to_parent_id_str];
        console.log('Working for', parent);
        while (taskqueue.length) {
            db.collection('tweets').findOne(
                {'id_str': parent},
                {},
            );
        }
    }
}


async function main() {
    const twitter_id = process.argv[2];

    var db = await connectToDB(env.mongodb, env.database);
    await prepareGraph(twitter_id, db);

    console.log(Object.keys(graph).length);

    process.exit(0);
}


if (typeof require != 'undefined' && require.main == module) {
    main();
}

File: metadata-learning/create-conversation-graph.py
#!/usr/bin/env python3

from pymongo import MongoClient
import argparse
import json
from concurrent.futures import ThreadPoolExecutor, as_completed


def parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument('-m', '--mongodb', help='mongodb server url',
                        default='')
    parser.add_argument('-d', '--dbname', help='mongodb database name',
                        default='summer-project')
    parser.add_argument('-t', '--twitter-id', help='twitter page to work for',
                        required=True)
    parser.add_argument('-o', '--outfile', help='resulting json file')
    parser.add_argument('-v', '--verbose', help='log everything',
                        action='store_false')
    return parser.parse_args()


def do_it_for_one(args, db, each):
    if args.verbose:
        print()
    # graph[each['id_str']] = [each]
    each['_id'] = str(each['_id'])
    result = [each]
    cur = each
    if args.verbose:
        print('Got', each['id_str'])
        print('adding', end=' ')
    while cur['in_reply_to_status_id_str']:
        try:
            # newone = next(db.tweets.find(
            #     {'id_str': cur['in_reply_to_status_id_str']},
            #     {'_id': True,
            #     'in_reply_to_status_id_str': True,
            #     'id_str': True})
            # )
            newone = next(db.tweets.find(
                {'id_str': cur['in_reply_to_status_id_str']},
            ))
            newone['_id'] = str(newone['_id'])
            result.append(newone)
            cur = newone
            if args.verbose:
                print(newone['id_str'], end=' ')
        except StopIteration:
            if args.verbose:
                print('Stop iteration hit')
            cur = {'in_reply_to_status_id_str': None}

    return result


def prepare_graph(args, db):
    graph = {}
    # org_tweets = db.tweets.find({'user.screen_name': args.twitter_id,
    #                              'in_reply_to_status_id_str': {'$ne': None}},
    #                             {'_id': True,
    #                              'in_reply_to_status_id_str': True,
    #                              'id_str': True})
    org_tweets = db.tweets.find({'user.screen_name': args.twitter_id,
                                 'in_reply_to_status_id_str': {'$ne': None}},
                                )

    with ThreadPoolExecutor(max_workers=10) as executor:
        futures = {executor.submit(do_it_for_one, args, db, each): each['id_str']
                   for each in org_tweets}
        for fut in as_completed(futures):
            graph[futures[fut]] = fut.result()

    with open(args.outfile, 'w') as outfile:
        outfile.write(json.dumps(graph))


def main():
    args = parse_args()
    if not args.outfile:
        args.outfile = args.twitter_id + '.json'

    if args.mongodb:
        db = MongoClient(args.mongodb).get_database(args.dbname)
    else:
        db = MongoClient().get_database(args.dbname)

    print(db, args)

    prepare_graph(args, db)


if __name__ == '__main__':
    main()


